### ISSUE TITLE
actions/force stalling and retrying

### ISSUE THREAD

**Issue by alexcrea** [VedalAI/neuro-sdk#14]
According to the documentation, Neuro should retry a force action if the success is false.
It would be nice to have an, optional, "no_retry" flag on Action Result message to avoid force action being replayed on the Neuro side if we think it does not apply anymore, or we want to manually handle force action retry.
Scenario example where it can be useful:

[Game] created an action, possible to be used but do not yet forced.
Things happens game the game...
[Neuro] try to use the action by herself
[Game] send a force action with only this action (did not yet receive the first neuro action request: communication ping)
[Neuro] received the force action, try to do the action again
[Game] received the first action. Unregister the action
[Game] received second action. It sees that the action is not registered and/or that the action is not possible anymore. So it sends an unsuccessful result. 
[Neuro] need to retry this force action as she received an unsuccessful Action Result
(behavior not specified in the API doc in this case, as the action used in force action got unregistered. So she needs to retry a force action with no actions ?)

Another similar scenario would be sending a disposable action and neuro try this action before the force action is received due to comically bad timing

The current workaround is sending a successful action result with an empty message.

(Also Randy, do not currently retry force action with unsuccessful action result.)

**Comment by alexcrea** [comment]
First, I'm sorry for not providing the scenario as an sequenceDiagram as I did not know it existed.
Secondly, I may have forgotten that it could be safe by Neuro being "busy"/"frozen" while the action is not yet sent. My bad. 
Also, maybe should document that action force is discarded when no valid (not registered) action (neuro side) is provided.

I'm obviously less competent in this domain, so good luck on finding a solution for the other potentials problems.

(also, could be good, and a lot less weird, to have an "is_forced"/"forced" field on Incoming Action message just to be sure, game side, that everything is ok. but probably not useful in most case so just ignore this at will.)

**Comment by Vedal987** [comment]
> > If what I understand is correct and the server drops the next action force because the actions got unregistered, then if Neuro receives a failure result she won't retry it because it wasn't part of an action force, and then the action force is dropped, leaving the game waiting for an action and Neuro not knowing she needs to send one

How can she be expected to have an actions/force with an action that isn't registered? This is expected behavior because the game client is deliberately unregistering the action. Don't unregister the action if it wasn't successful. And in that case the actions/force won't be dropped since the action wasn't unregistered.

> Writing your scenario more explicitly, I think it works like this to my understanding:

Under the hood it actually works slightly differently than what is described here. The actions/force won't be stalled and instead will be attempted in parallel but still dropped if the action is unregistered as above. More info:

After Neuro receives the actions/force I believe she will immediately start processing it while waiting for her other action result. However, this won't be committed to memory yet since we're still not sure if the action was successful and will lead to parallel processing for Neuro. Then there are two possibilities:

Neuro receives the actions/unregister BEFORE deciding to use that action again:
This will cause Neuro to use an invalid action and the actions/force will be retried which should cause the actions/force to be discarded if there are no more valid actions she can use.

Neuro receives the actions/unregister AFTER deciding to use that action again:
If the game client sends success as false then like before a retry will be attempted and discarded. If the game client sends success true then it's more than welcome to.

**Comment by KTrain5169** [comment]
I would like to see an update as to whether or not something like this is still being considered, as the current situation is potentially counterintuitive and confusing.

> Having a no_retry flag would be kind of weird because it would basically do nothing if the action wasn't part of an action force.

if that's the case, I propose that `no_retry` also be used outside action forces to determine whether or not to prompt neuro to more strongly consider retrying the action itself.

**Comment by Vedal987** [comment]
Current behaviour, now documented in the spec and [best practices](https://github.com/VedalAI/neuro-sdk/blob/main/API/BEST_PRACTICES.md): a `success: false` result on a forced action causes a limited number of automatic retries, after which the force is abandoned, and a retry whose actions have all been unregistered is discarded. There is also a timeout of about 30 seconds on waiting for a result, so a permanently stalled force shouldn't be possible.

Regarding the original question: sending `success: true` with an explanatory message is the proper way to say "this didn't work, but don't retry it" (as per the tip in the spec), so I'm not planning a separate `no_retry` flag.
